components/benefit_markets/app/domain/benefit_markets/operations/pvp/create_pvp_eligibility.rb

```diff
@@ -0,0 +1,121 @@
+    module Pvp
+      class CreatePvpEligibility
+        send(:include, Dry::Monads[:result, :do])
+
+        attr_accessor :subject
+
+        def call(params)
+          values = yield validate(params)
+          eligibility_record = yield find_eligibility
+          eligibility_options = yield build_eligibility_options(values, eligibility_record)
+          eligibility = yield create_eligibility(eligibility_options)
+          eligibility_record = yield store(eligibility)
+
+          Success(eligibility_record)
+        end
+
+        private
+
+        def validate(params)
+          errors = []
+          errors << "evidence key missing" unless params[:evidence_key]
+          errors << "evidence value missing" unless params[:evidence_value]
+          errors << "effective date missing" unless params[:effective_date].is_a?(::Date)
+
+          @subject = GlobalID::Locator.locate(params[:subject])
+          errors << "subject missing or not found for #{params[:subject]}" unless @subject.present?
+          errors << "current_user missing or not found for #{params[:current_user]}" unless params[:current_user].present?
+
+          errors.empty? ? Success(params) : Failure(errors)
+        end
+
+        def find_eligibility
+          eligibility = subject.eligibilities.by_key(:cca_shop_pvp_eligibility).max_by(&:created_at)
+
+          Success(eligibility)
+        end
+
+        def build_eligibility_options(values, eligibility_record = nil)
+          ::Operations::Eligible::BuildEligibility.new(
+            configuration:
+              ::BenefitMarkets::Operations::Pvp::PvpEligibilityConfiguration.new(
+                subject: subject,
+                effective_date: values[:effective_date]
+              )
+          ).call(
+            values.merge(
+              eligibility_record: eligibility_record,
+              evidence_configuration:
+                ::BenefitMarkets::Operations::Pvp::PvpEvidenceConfiguration.new
+            )
+          )
+        end
+
+        # Following Operation expects AcaEntities domain class as subject
+        def create_eligibility(eligibility_options)
+          ::AcaEntities::Eligible::AddEligibility.new.call(
+            subject: "AcaEntities::BenefitMarkets::PremiumValueProduct",
+            eligibility: eligibility_options
+          )
+        end
+
+        def store(eligibility)
+          eligibility_record = subject.eligibilities.where(id: eligibility._id).first
+
+          if eligibility_record
+            update_eligibility_record(eligibility_record, eligibility)
+          else
+            eligibility_record = create_eligibility_record(eligibility)
+            subject.eligibilities << eligibility_record
+          end
+
+          if subject.save
+            Success(eligibility_record.reload)
+          else
+            Failure(subject.errors.full_messages)
+          end
+        end
+
+        def update_eligibility_record(eligibility_record, eligibility)
```

Comment:
if i'm following this correctly, at this point we are sending in existing/old eligiblity, and new eligibility here and recording state change history for multiple objects? Might consider if this would be useful to indicate in a comment at the class level (and/or tweak param names here?).

Reply:
helpful - thanks!